clash-prelude/src/Clash/Class/Counter.hs

```diff
@@ -0,0 +1,151 @@
+{-|
+  Copyright   :  (C) 2021     , QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Utilities for wrapping counters consisting of multiple individual counters
+-}
+
+{-# LANGUAGE TypeFamilies #-}
+
+module Clash.Class.Counter where
+
+import Clash.Class.Num (SaturatingNum, SaturationMode(SatWrap), satSucc, satPred)
+import Clash.Sized.BitVector (BitVector)
+import Clash.Sized.Index (Index)
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.Unsigned (Unsigned)
+
+import GHC.TypeLits (KnownNat, type (<=))
+
+-- | 'Counter' is a class that composes multiple counters into a single one. For
+-- example, it allows you to construct a counter from multiple smaller counters.
+-- It is similar to odometers found in olds cars, once all counters reach their
+-- maximum they reset to zero - i.e. odometer rollover. See 'countSucc' and
+-- 'countPred' for API usage examples.
+--
+-- Example use case: when driving a monitor through VGA you would like to keep
+-- track at least two counters: one counting a horizontal position, and one
+-- vertical. Perhaps a fancy VGA driver would also like to keep track of the
+-- number of drawn frames. To do so, the three counters are setup with different
+-- types. On each _round_ of the horizontal counter the vertical counter should
+-- be increased. On each _round_ of the vertical counter the frame counter should
+-- be increased. With this class you could simply use the type:
+--
+--   (FrameCount, VerticalCount, HorizontalCount)
+--
+-- and have 'countSucc' work as described.
+--
+class (Eq a, Bounded a) => Counter a where
+  -- | Successor of a counter
+  --
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (1, 1)
+  -- (1, 2)
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (1, 2)
+  -- (1, 3)
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (1, 3)
+  -- (2, 0)
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (3, 3)
+  -- (0, 0)
+  countSucc :: a -> a
+  default countSucc :: SaturatingNum a => a -> a
+  countSucc = satSucc SatWrap
+
+  -- | Predecessor of a counter
+  --
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (1, 2)
+  -- (1, 1)
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (1, 3)
+  -- (1, 2)
+  -- >>> countSucc @(Unsigned 2, Unsigned 2) (2, 0)
+  -- (1, 3)
+  -- >>> countPred @(Unsigned 2, Unsigned 2) (0, 0)
+  -- (3, 3)
+  countPred :: a -> a
+  default countPred :: SaturatingNum a => a -> a
+  countPred = satPred SatWrap
+
+instance (1 <= n, KnownNat n) => Counter (Index n)
+instance KnownNat n => Counter (Unsigned n)
+instance KnownNat n => Counter (Signed n)
+instance KnownNat n => Counter (BitVector n)
+
+instance (Counter a, Counter b) => Counter (a, b) where
```

Comment:
All these tuple instances seem pretty mechanical, do you think it would be worth using template Haskell?

Reply:
Yes! It's even a TODO in that file :)

Reply:
If only Haskell were like BlueSpec, and any n-tuple larger than a 2-tuple would be desugared to a nested 2-tuple, meaning you only ever have to define a 2-tuple instances.